File: Backend/services/pineconeService.js
/**
 * Pinecone Service - Store and retrieve RFP documents for chatbot
 */
const { Pinecone } = require('@pinecone-database/pinecone');
const { Document } = require('@langchain/core/documents');
const { RecursiveCharacterTextSplitter } = require('@langchain/textsplitters');
const { PineconeStore } = require('@langchain/pinecone');

// Initialize Pinecone
let pineconeClient = null;
let pineconeIndex = null;

const initializePinecone = () => {
    if (!pineconeClient && process.env.PINECONE_API_KEY) {
        pineconeClient = new Pinecone({
            apiKey: process.env.PINECONE_API_KEY
        });
    }
    return pineconeClient;
};

/**
 * Store RFP document in Pinecone for chatbot queries
 * @param {string} documentId - Unique ID for the document (e.g., file hash)
 * @param {string} fileName - Original filename
 * @param {string} text - Extracted text from document
 * @param {Object} metadata - Additional metadata (analysis results, etc.)
 */
const storeRFPInPinecone = async (documentId, fileName, text, metadata = {}) => {
    try {
        const client = initializePinecone();
        if (!client) {
            throw new Error('Pinecone not initialized. Check PINECONE_API_KEY.');
        }

        const indexName = process.env.PINECONE_INDEX_NAME || 'bid-intelligence-chatbot';
        
        // Get or create index (support Host URL if provided)
        const indexHostUrl = process.env.PINECONE_INDEX_HOST;
        // The Pinecone v3 client uses lowercase .index()
        const index = client.index(indexName, indexHostUrl ? indexHostUrl : undefined);

        // Split text into chunks
        const textSplitter = new RecursiveCharacterTextSplitter({
            chunkSize: 1000,
            chunkOverlap: 200,
        });

        const chunks = await textSplitter.splitText(text);
        
        // Create documents with metadata
        const documents = chunks.map((chunk, index) => {
            return new Document({
                pageContent: chunk,
                metadata: {
                    documentId,
                    fileName,
                    chunkIndex: index,
                    totalChunks: chunks.length,
                    uploadedAt: new Date().toISOString(),
                    ...metadata
                }
            });
        });

        // Store in Pinecone using LangChain
        const vectorStore = await PineconeStore.fromDocuments(
            documents,
            // Note: You'll need to use the same embeddings model as the chatbot
            // For now, we'll use a placeholder - you may need to adjust this
            await getEmbeddings(),
            {
                pineconeIndex: index,
                namespace: documentId, // Use documentId as namespace for easy retrieval
            }
        );

        console.log(`✅ Stored ${chunks.length} chunks for document: ${fileName} (ID: ${documentId})`);
        
        return {
            success: true,
            documentId,
            chunksCount: chunks.length,
            indexName
        };

    } catch (error) {
        console.error('Error storing document in Pinecone:', error);
        throw error;
    }
};

/**
 * Get embeddings - using the same model as chatbot
 * This should match the embeddings used in the Python chatbot
 */
const getEmbeddings = async () => {
    // For now, we'll need to call the Python service or use a compatible embedding model
    // Since the chatbot uses sentence-transformers, we might need to use a compatible Node.js library
    // or make an API call to the Python service
    
    // Option 1: Use OpenAI embeddings (compatible)
    const { OpenAIEmbeddings } = require('@langchain/openai');
    return new OpenAIEmbeddings({
        openAIApiKey: process.env.OPENAI_API_KEY,
        modelName: 'text-embedding-3-small' // or 'text-embedding-ada-002'
    });
    
    // Option 2: If you want to use the same model as Python (sentence-transformers),
    // you would need to make an HTTP call to a Python service or use a compatible library
};

/**
 * Query documents by documentId
 * @param {string} documentId - Document ID to query
 * @param {string} query - Search query
 * @param {number} k - Number of results
 */
const queryRFPDocument = async (documentId, query, k = 3) => {
    try {
        const client = initializePinecone();
        if (!client) {
            throw new Error('Pinecone not initialized.');
        }

        const indexName = process.env.PINECONE_INDEX_NAME || 'bid-intelligence-chatbot';
        const indexHostUrl = process.env.PINECONE_INDEX_HOST;
        const index = client.index(indexName, indexHostUrl ? indexHostUrl : undefined);

        const embeddings = await getEmbeddings();
        
        const vectorStore = await PineconeStore.fromExistingIndex(embeddings, {
            pineconeIndex: index,
            namespace: documentId,
        });

        const results = await vectorStore.similaritySearch(query, k);
        
        return results;

    } catch (error) {
        console.error('Error querying document:', error);
        throw error;
    }
};

/**
 * Delete document from Pinecone
 * @param {string} documentId - Document ID to delete
 */
const deleteRFPFromPinecone = async (documentId) => {
    try {
        const client = initializePinecone();
        if (!client) {
            throw new Error('Pinecone not initialized.');
        }

        const indexName = process.env.PINECONE_INDEX_NAME || 'bid-intelligence-chatbot';
        const indexHostUrl = process.env.PINECONE_INDEX_HOST;
        const index = client.index(indexName, indexHostUrl ? indexHostUrl : undefined);

        // Delete all vectors in the namespace
        await index.deleteMany({
            filter: { documentId: { $eq: documentId } }
        });

        console.log(`✅ Deleted document from Pinecone: ${documentId}`);
        return { success: true };

    } catch (error) {
        console.error('Error deleting document:', error);
        throw error;
    }
};

module.exports = {
    storeRFPInPinecone,
    queryRFPDocument,
    deleteRFPFromPinecone,
    initializePinecone
};

File: Backend_py/services/pinecone_service.py
import os
import logging
from typing import List, Dict, Any, Optional
from core.config import settings

logger = logging.getLogger(__name__)

# Optional imports for Pinecone - fail gracefully if not available
try:
    from pinecone import Pinecone
    from langchain_pinecone import PineconeVectorStore
    from langchain_core.documents import Document as LCDocument
    from langchain_community.embeddings import HuggingFaceEmbeddings
    # Try different import paths for text splitter depending on langchain version
    try:
        from langchain.text_splitters import RecursiveCharacterTextSplitter
    except ImportError:
        try:
            from langchain.text_splitter import RecursiveCharacterTextSplitter
        except ImportError:
            from langchain_community.text_splitter import RecursiveCharacterTextSplitter
    PINECONE_AVAILABLE = True
except ImportError as e:
    logger.warning(f"Pinecone dependencies not available: {e}. Pinecone features will be disabled.")
    PINECONE_AVAILABLE = False
    Pinecone = None
    PineconeVectorStore = None
    RecursiveCharacterTextSplitter = None
    LCDocument = None
    HuggingFaceEmbeddings = None

def get_embeddings():
    if not PINECONE_AVAILABLE or HuggingFaceEmbeddings is None:
        raise ImportError("HuggingFaceEmbeddings not available. Pinecone dependencies are missing.")
    return HuggingFaceEmbeddings(model_name='sentence-transformers/all-MiniLM-L6-v2')

def get_pinecone_client():
    if not PINECONE_AVAILABLE:
        return None
    if settings.PINECONE_API_KEY:
        return Pinecone(api_key=settings.PINECONE_API_KEY)
    return None

async def store_rfp_in_pinecone(document_id: str, file_name: str, text: str, metadata: Dict[str, Any] = None):
    if not PINECONE_AVAILABLE:
        logger.warning("Pinecone service not available. Skipping storage.")
        return None
    try:
        pc = get_pinecone_client()
        if not pc:
            logger.warning("Pinecone API key not set. Skipping storage.")
            return None
            
        index_name = os.getenv("PINECONE_INDEX_NAME", "bid-intelligence-chatbot")
        index = pc.Index(index_name)
        
        splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
        chunks = splitter.split_text(text)
        
        docs = []
        for i, chunk in enumerate(chunks):
            meta = {
                "documentId": document_id,
                "fileName": file_name,
                "chunkIndex": i,
                "totalChunks": len(chunks)
            }
            if metadata:
                meta.update(metadata)
            docs.append(LCDocument(page_content=chunk, metadata=meta))
            
        embeddings = get_embeddings()
        
        vector_store = PineconeVectorStore.from_documents(
            docs,
            embeddings,
            index_name=index_name,
            namespace=document_id
        )
        
        logger.info(f"✅ Stored {len(chunks)} chunks in Pinecone for {file_name} in namespace {document_id}")
        return {"success": True, "chunks": len(chunks)}
    except Exception as e:
        logger.error(f"Error storing in Pinecone: {str(e)}")
        raise e

async def query_rfp_document(document_id: str, query: str, k: int = 3):
    if not PINECONE_AVAILABLE:
        logger.warning("Pinecone service not available. Cannot query.")
        return []
    try:
        embeddings = get_embeddings()
        index_name = os.getenv("PINECONE_INDEX_NAME", "bid-intelligence-chatbot")
        
        vector_store = PineconeVectorStore.from_existing_index(
            index_name=index_name,
            embedding=embeddings,
            namespace=document_id
        )
        
        results = vector_store.similarity_search(query, k=k)
        return results
    except Exception as e:
        logger.error(f"Error querying Pinecone: {str(e)}")
        return []
